Oss select API

qixu001 opened the pull request.

Commit:
add sign-v2 code and doc.

Commit:
add PostObject V2 signature example

Commit:
fix syntax error in Python 2.6; skip V2 only cases when V1 sign is
specified.

Commit:
fix readme for sign v2 testing

Commit:
fix #82, #73

Commit:
fix #87

Commit:
fix ut error in py3 due to using xrange

baiyubin2020 committed:
Merge pull request #85 from aliyun/fix-issues

Fixed #82 #73 #87

baiyubin2020 committed:
fix #64

baiyubin2020 committed:
fix #64

Commit:
Merge pull request #89 from aliyun/fix-issues-64

fix #64

baiyubin2020 committed:
release 2.3.4

baiyubin2020 committed:
Merge pull request #90 from aliyun/fix-issues-64

release 2.3.4

wujinhu committed:
create_bucket supports IA/Archive
add restore_object

wujinhu committed:
remove unused import

wujinhu committed:
add get_bucket_stat

wujinhu committed:
add get_bucket_info

wujinhu committed:
lifecycle rule

wujinhu committed:
lifecycle rule

wujinhu committed:
lifecycle rule

wujinhu committed:
fix ut

wujinhu committed:
fix ut

wujinhu committed:
add ut

wujinhu committed:
add ut

wujinhu committed:
add ut

wujinhu committed:
add ut

wujinhu committed:
add ut

wujinhu committed:
add ut

wujinhu committed:
add ut

wujinhu committed:
add ut

wujinhu committed:
add ut

wujinhu committed:
add ut

wujinhu committed:
fix syntax error

wujinhu committed:
fix ut

wujinhu committed:
optimize ut

wujinhu committed:
restore ut

wujinhu committed:
restore ut

wujinhu committed:
restore ut

wujinhu committed:
optimize ut

wujinhu committed:
optimize ut

wujinhu committed:
optimize ut

wujinhu committed:
optimize ut

wujinhu committed:
some code fix

wujinhu committed:
some code fix

wujinhu committed:
some code fix

wujinhu committed:
some code fix

Commit:
Merge pull request #94 from aliyun/createBucket

create_bucket supports IA/Archive && add restore_object

wujinhu committed:
Merge branch 'master' into sign-v2

# Conflicts:
#	README.rst

baiyubin2020 committed:
Release 2.4.0

wujinhu committed:
refactor auth

baiyubin2020 committed:
Release 2.4.0

baiyubin2020 committed:
Merge pull request #95 from aliyun/createBucket

Release 2.4.0

wujinhu committed:
translations

wujinhu committed:
translations

wujinhu committed:
translations

wujinhu committed:
optimize some code

wujinhu committed:
optimize some code

wujinhu committed:
optimize some code

wujinhu committed:
optimize some code

wujinhu committed:
test

wujinhu committed:
test

wujinhu committed:
test

wujinhu committed:
test

wujinhu committed:
test

wujinhu committed:
test

wujinhu committed:
test

wujinhu committed:
test

wujinhu committed:
test

wujinhu committed:
test

wujinhu committed:
test

wujinhu committed:
fix name

wujinhu committed:
revert some code

wujinhu committed:
revert some code

wujinhu committed:
revert some code

jubel-han committed:
support the get object with specified query parameters

Commit:
Merge pull request #96 from aliyun/sign-v2

OSS Sign v2 support

jubel-han committed:
add the tests for query parameters

mingyangyang committed:
add client crypto support(local rsa) for basic put and get object operations

mingyangyang committed:
use local variable to store key passphrase, add test cases

mingyangyang committed:
add kms cipher mode

jubel-han committed:
adjust the query parameter test cases

mingyangyang committed:
fix test case for py3.3 and py3.4, increase travis waiting time

mingyangyang committed:
get more log in travis running

Commit:
Merge pull request #99 from jubel-han/master

Supporting with query string parameter when getting objects

mingyangyang committed:
modify travis test env

mingyangyang committed:
fix pr#99 test case

mingyangyang committed:
Merge branch 'master' into crypto-dev

mingyangyang committed:
code optimize

mingyangyang committed:
add some unittest case after review

mingyangyang committed:
provide CryptoBucket for client-encryption, only support put_object and get_object

mingyangyang committed:
disable test_kms_with_error_parameter in py3, because aliyun-python-sdk-core-v3 could not handle error response

mingyangyang committed:
optimize code

qixu001 committed:
add oss-select APIs

three APIs:
select_object: select object content to memory
select_object_to_file:select object to file
head_csv_object: head csv object

qixu001 committed:
rename the select.py to mitigate the confliction with system module select

qixu001 committed:
rename select API并增加说明

qixu001 committed:
添加select FT

qixu001 committed:
update response data parser accordingly for the  OSS end frame format change

1) The end frame's checksum bytes are now in the end of the frame (after the status and error message)
2) also remove the use of make_progress_adapter cause the SelectFrameResponse includes the progress update function

qixu001 committed:
add total_bytes_to_scan attribute into SelectObjectResult

The attribute is used for progress update

qixu001 committed:
hook the progress update on select response's iterator function

qixu001 committed:
add more select test cases

qixu001 committed:
update the python sdk according to match the server side change

The x-oss-select-total-bytes-to-scan is removed in server's behavior and thus the SDK needs the change as well.

qixu001 committed:
add the split based range query for oss-select API

mingyangyang committed:
optimize code, add custom_crypto.py

qixu001 committed:
update the python SDK to use POST body for parameters

qixu001 committed:
fix the cycling reference issue between api and xml_utils

qixu001 committed:
delete the unused code

baiyubin commented on examples/select_csv.py:
```diff
@@ -0,0 +1,55 @@
+                'RecordDelimiter': '\r\n',
+                'LineRange': (500, 1000)}
+
+csv_header = bucket.get_csv_object_meta(key, csv_meta_params)
```
- get_csv_object_meta, select_csv_object, select_csv_object_to_file 名字是以cvs视角，是不是应该以select为视角，叫 get_select_object_meta, select_object, select_object_to_file。Java SDK是采用了后者，多个sdk最好统一一下。

qixu001 replied:
用CSV作为名字是因为以后如果其他文件类型它的实现是不太一样的。不过这个用统一API名字，在实现里再做区分也可以。那就和Java SDK统一一下吧

baiyubin commented on oss2/api.py:
```diff
@@ -108,6 +133,28 @@ def progress_callback(bytes_consumed, total_bytes):
+.. _select_csv_params:
+
+    指定OSS Select的CSV文件格式，支持如下Keys:
+    >>> FileHeaderInfo: None|Use|Ignore   #None表示没有CSV Schema头，Use表示启用CSV Schema头，可以在Select语句中使用Name，Ignore表示有CSV Schema头，但忽略它（Select语句中不可以使用Name)
```
- FileHeaderInfo 是不是叫 CvsHeaderInfo 更能说明语义

qixu001 replied:
嗯，这个是模仿AWS的，它也叫FileHeaderInfo。不过这个在SDK里可以换个更贴切的名字。服务器端还是叫FileHeaderInfo.

baiyubin commented on oss2/api.py:
```diff
@@ -514,6 +613,40 @@ def head_object(self, key, headers=None):
         resp = self.__do_object('HEAD', key, headers=headers)
         return HeadObjectResult(resp)
     
+    def get_csv_object_meta(self, key, csv_meta_params=None):
```
- get_csv_object_meta 的实际操作是产生Meta，应该叫 create_select_meta 或者 generate_select_meta

qixu001 replied:
这个我觉得叫get_csv_object_meta更为妥当。因为以后我们很有可能会自动产生这个meta，那个时候就真的变成get了。而且，这里如果meta已经存在是不会再重复create的，所以大概率还是get。

baiyubin commented on oss2/api.py:
```diff
@@ -514,6 +613,40 @@ def head_object(self, key, headers=None):
+
+        self.timeout = 3600
+        resp = self.__do_object('POST', key, data = body, headers=headers, params=params)
+        return HeadCsvObjectResult(resp)
```
- HeadCsvObjectResult 应该和函数名称同步修改

qixu001 replied:
ok

baiyubin commented on oss2/models.py:
```diff
@@ -75,6 +76,11 @@ def __init__(self, resp):
+    def __init__(self, resp):
+        super(HeadCsvObjectResult, self).__init__(resp)
+        self.CsvRows = int(self.headers['x-oss-select-csv-rows'])
+        self.CsvSplits = int(self.headers['x-oss-select-csv-splits'])
```
- CsvRows, CsvSplits ->  csv_rows, csv_splits

qixu001 replied:
ok

baiyubin commented on oss2/select_response.py:
```diff
@@ -0,0 +1,164 @@
+_CONTINIOUS_FRAME_TYPE=8388612
+_DATA_FRAME_TYPE = 8388609
+_END_FRAME_TYPE = 8388613
+_FRAMES_FOR_PROGRESS_UPDATE = 10
```
- 上面的常量只有SelectFrameResponse用到，并且是不希望用户修改的，请定义到类SelectFrameResponse里面

qixu001 replied:
ok

baiyubin commented on oss2/select_response.py:
```diff
@@ -0,0 +1,164 @@
+_DATA_FRAME_TYPE = 8388609
+_END_FRAME_TYPE = 8388613
+_FRAMES_FOR_PROGRESS_UPDATE = 10
+class SelectFrameResponse(object):
```
- SelectFrameResponse -> _SelectResponseAdapter，格式加注释说明

qixu001 replied:
ok

baiyubin commented on oss2/xml_utils.py:
```diff
@@ -492,3 +493,41 @@ def to_create_live_channel(live_channel):
+            _add_text_child(csv, 'Range', utils._make_line_range_string(select_csv_params['LineRange']))
+
+    return _node_to_string(root)
+def to_create_head_body(csv_meta_param):
```
- to_create_head_body 名称根据方法修改
- 前面加空行

qixu001 replied:
ok

baiyubin commented on tests/test_select_csv_object.py:
```diff
@@ -0,0 +1,209 @@
+        testCase.assertGreater(len(content), 0)
+        if line_range is None:
+            testCase.assertEqual(self.scannedSize, file_size)
+        return content
```
- 这部分测试的代码写的感觉比较仓促，请优化一下，空行/空行都正确使用一下，使代码更容易理解。

qixu001 replied:
ok

qixu001 committed:
address code review comments

1) rename select API name to select_object,
select_object_to_file,
get_select_object_meta.
2) Rename FileHeaderInfo with CsvHeaderInfo
3) Rename HeadCsvObjectResult with
4) rename CsvRow, CsvSplit with csv_row, csv_split
5) move constants to class SelectResponseAdapter
6) rename SelectFrameResponse with SelectResponseAdapter and add comments on the format

qixu001 committed:
rename get_select_object_meta to create_select_object_meta

qixu001 committed:
fix the UT failures
remove unneeded code

qixu001 committed:
Merge branch 'oss-select-preview' into oss-select

qixu001 committed:
Merge pull request #114 from qixu001/oss-select

Oss select APIs for preview

baiyubin committed:
Merge pull request #103 from aliyun/crypto-dev

add client crypto support for basic put and get object ope…

baiyubin2020 committed:
release 2.5.0

qixu001 committed:
add gzip support and some minor fixes

yangwanyuan committed:
add check_crc to get_object_to_file

yangwanyuan committed:
add crc64_combine.py to support crc64 combine feature

yangwanyuan committed:
add crc64_combine and check_crc to Multipart Upload

yangwanyuan committed:
add check_crc to resumable download

yangwanyuan committed:
fix test and unittest error compare to release 2.5.0

yangwanyuan committed:
add crc related tests and examples

yangwanyuan committed:
fix code to make it compatiable with python3

qixu001 committed:
update the python SDK for select

hangzws committed:
add detail logs

hangzws committed:
add thread id in all logs & add logs for init auth

yangwanyuan committed:
add some check crc functional test

hangzws committed:
add extranet_endpoint&intranet_endpoint&storage_class in list buckets result

hangzws committed:
quote source_key in upload part copy

hangzws committed:
add support for ipv6

yangwanyuan committed:
add requestHeader

yangwanyuan committed:
add requestHeader unittest

yangwanyuan committed:
add server crypto example

yangwanyuan committed:
fix setServerSideEncryto bug

yangwanyuan committed:
replace some header string with variable in headers.py

hangzws committed:
add process object & put/get object with signed url

hangzws committed:
add api for put/get with signed url and tests

qixu001 committed:
update the head API's response handler

yangwanyuan committed:
Merge branch 'log' into ywy-crc-dev

yangwanyuan committed:
Merge branch 'log' into ywy-cmkid-dev

hangzws committed:
add test case for process object

hangzws committed:
use bytes as base64 encode param

hangzws committed:
transfer encode base64 bytes to string

hangzws committed:
add comment for process object

qixu001 committed:
refine the code to address comments

1) change the endianness only if the machine is little endian
2) add client side select exception
3) use crcmod for crc32 computation
4) remove the commented code
5) refine the status check code

qixu001 committed:
fix a compile issue

yangwanyuan committed:
fix code after review

yangwanyuan committed:
fix code after review

yangwanyuan committed:
fix unittests in python 3

qixu001 committed:
add select ut

add select ut

qixu001 committed:
renaming the test helper class name

qixu001 committed:
add UT fix for python 3.5

qixu001 committed:
fix the UT for python2.6

yangwanyuan committed:
code review fix

yangwanyuan committed:
add crc64 combine unittest

hangzws committed:
Merge pull request #121 from qixu001/oss-select

Oss select

qixu001 committed:
fix a ft regression

hangzws committed:
Merge pull request #128 from qixu001/oss-select

fix a ft regression

qixu001 committed:
add error message in the response of API create_select_object_meta

The oss server side just added an error message field in the response body of create_select_object_meta.

yangwanyuan committed:
remove note in crc64_combine, remove request_id from download

hangzws committed:
Merge pull request #125 from yangwanyuan/ywy-crc-dev

Ywy crc dev

hangzws committed:
Merge branch 'log' into ywy-cmkid-dev

hangzws committed:
Merge pull request #126 from yangwanyuan/ywy-cmkid-dev

Ywy cmkid dev

hangzws committed:
Merge pull request #129 from qixu001/oss-select

add error message in the response of API create_select_object_meta

hangzws committed:
Merge branch 'log' of github.com:aliyun/aliyun-oss-python-sdk into log

hangzws committed:
variable&func naming specification modification

hangzws committed:
add ut for init

hangzws committed:
fix process object return & move ut to ft

hangzws committed:
convert http response to string while py3

hangzws committed:
merge master to select_preview

hangzws committed:
Merge pull request #130 from aliyun/log

Log

hangzws committed:
change endpoint for select ft

hangzws committed:
fix compatibility issues in test case

hangzws committed:
remove unused select_date and fix print error in py3

hangzws committed:
merge master changes to select_preview

hangzws committed:
only test v1 for timeout reason

hangzws committed:
Merge pull request #131 from aliyun/oss-select-preview

Oss select preview

hangzws committed:
release 2.6.0

qixu001 committed:
Merge pull request #2 from aliyun/master

sync from aliyun's official version

qixu001 committed:
add json file support for OSS select APIs
